internal/controllers/core/extensionrepo/reconciler.go

```diff
@@ -40,9 +65,117 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	}
 
 	if apierrors.IsNotFound(err) || !repo.ObjectMeta.DeletionTimestamp.IsZero() {
-		// TODO(nick): Handle deletion
+		// If a repo is deleted, delete it on disk.
+		//
+		// This is simple, but not accurate.
+		// 1) It will garbage collect two aggressively if two extension repo
+		//    objects point to the same URL.
+		// 2) If we "miss" a delete event, the repo will never get cleaned up.
+		//
+		// A "real" implementation would use the on-disk repos as the source of
+		// truth, and garbage collect ones with no remaining refs.
+		path, ok := r.lastKnownDestPaths[nn]
+		if !ok {
+			return ctrl.Result{}, nil
+		}
+		err := os.RemoveAll(path)
+		if err != nil && !os.IsNotExist(err) {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Check that the URL is valid.
+	importPath, err := getImportPath(&repo)
+	if err != nil {
+		logger.Get(ctx).Errorf("Invalid ExtensionRepo %s: %v", repo.Name, err)
+		return r.updateStatus(ctx, &repo, func(status *v1alpha1.ExtensionRepoStatus) {
+			status.Error = err.Error()
+		})
+	}
+
+	dlr := get.NewDownloader(r.dlrPath)
+	destPath := dlr.DestinationPath(importPath)
+	info, err := os.Stat(destPath)
+	if err != nil && !os.IsNotExist(err) {
+		return ctrl.Result{}, err
+	}
+
+	// If the destination path does not exist, download it now.
+	if os.IsNotExist(err) {
+		lastFetch := r.fetches[nn]
+		lastBackoff := r.backoffs[nn]
+		if time.Since(lastFetch) < lastBackoff {
+			return ctrl.Result{RequeueAfter: lastBackoff}, nil
```

Comment:
should this be `lastBackoff - time.Since(lastFetch)`? or maybe it can just skip the RequeueAfter since there's still theoretically a requeued request coming in from the call that set the backoff in the first place? I guess this probably doesn't matter in practice.